### DISCUSSION TITLE
One-To-Many collection with a lot of additional null objects

### DISCUSSION THREAD

**Discussion by paradox2707** [nhibernate/nhibernate-core#3272] [Q&A] [answered]
I upgraded NHibernate from 3.3.5 to 5.3.14
I got a new limitation for left outer join with the error "Cannot simultaneously fetch multiple bags"
Hence, I decided to move from Bag to List collection type, using Fluent NHibernate like `HasMany(x => x.RelatedEntity).KeyColumn("MainEntityId").**AsList(x => x.Column("RelatedEntityId"))**` (fold code is newly added)
As a result, I get a collection of RelatedEntity with a lot of additional Null objects including correct instances
It is regarding any query over, with limitation of only 1 bag left outer join or not, so if I turn back to Bag type all work well, but a couple of query throw the error "Cannot simultaneously fetch multiple bags" again))))
Thanks in advance!

**Comment by pasqualedante** [comment, marked as answer]
This happens because the indexes of the elements of the list are not sequential, but there are gaps: if your index column has gaps in it (i.e. it has values like  3, 4, 5), NH would populate the resulting List with empty elements at the expected places (0,1 and 2). This behavior is by design as you can read at paragraph 6.3.2 of NH reference guide:

_6.2.3. Indexed collections
All collection mappings, except those with set and bag semantics, need an index column in the collection table.
An index column is a column that maps to an array index, or IList index, or IDictionary key. The index of an
IDictionary may be of any basic type, mapped with <map-key>. It can be an entity reference mapped with
<map-key-many-to-many> (or in exotic cases <index-many-to-any>, similar to the <any> mapping element), or
it can be a composite type mapped with <composite-map-key>. The index of an array or list is always of type
Int32 and is mapped using the <list-index> element. The mapped column contains sequential integers that are
numbered from zero by default._

To avoid _Cannot simultaneously fetch multiple bags_ you can map with unordered Set.

**Reply by fredericDelaporte** [reply]
Thanks,
Here is the link to the quoted documentation: https://nhibernate.info/doc/nhibernate-reference/collections.html#collections-indexed
